File: ui/review.js
// Review view: the diff of what the commit will contain, with line comments
// that end up in the deny reason. Loaded before app.js, which defines the
// shared helpers it calls at click time.

let files = [];        // FileDiff[] from the binary, each with a flat line list
let comments = [];     // { file, start, end, text, box }; start = FILE for a file comment
let drag = null;       // { file, start, end } while lines are being selected
let reviewOpen = false;
const rows = new Map(); // "file:index" -> line <tr>
const boxes = [];       // file index -> .file element
const checks = [];      // file index -> Viewed checkbox
const treeItems = [];   // file index -> <li> in the tree
const viewed = new Set();

const FILE = -1;
const MARK = { context: " ", add: "+", del: "-" };
const SCOPE_LABEL = {
  staged: "Staged changes only: plain git commit",
  tracked: "Tracked files as they are: git commit -a",
  worktree: "Working tree, untracked files included: git add runs first",
};
const SVG = "http://www.w3.org/2000/svg";
const CHEVRON = "M12.78 5.22a.749.749 0 0 1 0 1.06l-4.25 4.25a.749.749 0 0 1-1.06 0L3.22 6.28a.749.749 0 1 1 1.06-1.06L8 8.939l3.72-3.719a.749.749 0 0 1 1.06 0Z";
const BUBBLE = "M1 2.75C1 1.784 1.784 1 2.75 1h10.5c.966 0 1.75.784 1.75 1.75v7.5A1.75 1.75 0 0 1 13.25 12H9.06l-2.573 2.573A1.458 1.458 0 0 1 4 13.543V12H2.75A1.75 1.75 0 0 1 1 10.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h4.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z";

const el = (tag, className, text) => {
  const e = document.createElement(tag);
  if (className) e.className = className;
  if (text != null) e.textContent = text;
  return e;
};
const button = (label, onClick, className) => {
  const b = el("button", className, label);
  b.type = "button";
  b.onclick = onClick;
  return b;
};
const icon = (path) => {
  const svg = document.createElementNS(SVG, "svg");
  svg.setAttribute("viewBox", "0 0 16 16");
  svg.setAttribute("width", "16");
  svg.setAttribute("height", "16");
  const p = document.createElementNS(SVG, "path");
  p.setAttribute("d", path);
  p.setAttribute("fill", "currentColor");
  svg.append(p);
  return svg;
};

async function toggleReview() {
  reviewOpen = !reviewOpen;
  $("summary").hidden = reviewOpen;
  $("review").hidden = !reviewOpen;
  $("review-head").hidden = !reviewOpen;
  $("toggle-review").textContent = reviewOpen ? "Summary" : "Review changes";
  if (reviewOpen && !files.length) await loadReview();
}

async function loadReview() {
  $("files").textContent = "Reading the diff...";
  try {
    files = await invoke("changes");
  } catch (e) {
    $("files").textContent = "git error: " + e;
    return;
  }
  files.forEach((f) => { f.flat = f.hunks.flatMap((h) => h.lines); });
  $("review-scope").textContent = SCOPE_LABEL[scope] || "";
  $("files").textContent = "";
  $("files").classList.toggle("empty", !files.length);
  if (!files.length) $("files").textContent = "(nothing to commit in this scope)";
  files.forEach((f, i) => $("files").append(renderFile(f, i)));
  renderTree();
  files.forEach((f, i) => {
    if (f.viewed) setViewed(i, true);
    f.restored.forEach((r) => restore(i, r));
  });
  updateViewed();
  markCommented();
}

function setViewed(fi, on) {
  checks[fi].checked = on;
  if (on) viewed.add(fi); else viewed.delete(fi);
  boxes[fi].classList.toggle("collapsed", on);
  treeItems[fi].classList.toggle("viewed", on);
  updateViewed();
}

// A comment from an earlier attempt: back in place while its lines are
// unchanged, otherwise shown as outdated and not sent unless reopened.
function restore(fi, r) {
  if (r.anchor === "outdated") {
    renderOutdated(fi, r);
    return;
  }
  const c = r.anchor === "lines"
    ? { file: fi, start: r.start, end: r.end, text: r.text, earlier: true }
    : { file: fi, start: FILE, end: FILE, text: r.text, earlier: true };
  comments.push(c);
  renderComment(c);
}

function renderOutdated(fi, r) {
  const box = el("div", "comment outdated");
  const head = el("div", "comment-head");
  head.append(el("span", "author", user), el("span", "tag outdated", "Outdated"), el("span", "where", files[fi].path));
  box.append(head);
  if (r.quote.length) box.append(el("pre", "quote", r.quote.join("\n")));
  const actions = el("div", "comment-actions");
  actions.append(
    button("Dismiss", () => box.remove()),
    button("Reopen", () => {
      box.remove();
      const c = { file: fi, start: FILE, end: FILE, text: r.text };
      comments.push(c);
      renderComment(c);
      markCommented();
    }),
  );
  box.append(el("div", "comment-body", r.text), actions);
  commentCell(fi, FILE).append(box);
}

// What to keep for the next attempt; null when the diff was never opened.
function reviewState() {
  if (!files.length) return null;
  return files.map((f, i) => ({
    path: f.path,
    viewed: viewed.has(i),
    comments: comments.filter((c) => c.file === i).map((c) => ({
      start: c.start === FILE ? null : c.start,
      end: c.end === FILE ? null : c.end,
      text: c.text,
    })),
  }));
}

function renderFile(file, fi) {
  const box = el("div", "file");
  boxes[fi] = box;
  const head = el("div", "file-head");
  const chevron = button(null, () => box.classList.toggle("collapsed"), "chevron");
  chevron.append(icon(CHEVRON));
  chevron.title = "Collapse or expand";
  const added = file.flat.filter((l) => l.kind === "add").length;
  const removed = file.flat.filter((l) => l.kind === "del").length;
  const counts = el("span", "file-counts");
  counts.append(el("span", "add", "+" + added), " ", el("span", "del", "-" + removed));
  const viewedLabel = el("label", "viewed-label");
  const check = el("input");
  check.type = "checkbox";
  check.onchange = () => setViewed(fi, check.checked);
  checks[fi] = check;
  viewedLabel.append(check, "Viewed");
  const comment = button(null, () => openForm(fi, FILE, FILE), "file-comment");
  comment.append(icon(BUBBLE));
  comment.title = "Comment on this file";
  head.append(
    chevron,
    el("span", "file-status " + file.status, file.status),
    el("span", "file-path", file.old_path ? file.old_path + " -> " + file.path : file.path),
    counts,
    viewedLabel,
    comment,
  );
  box.append(head, el("div", "file-comments"));
  if (file.binary) {
    box.append(el("div", "file-note", "Binary file, no diff."));
    return box;
  }
  const table = el("table", "diff");
  let idx = 0;
  file.hunks.forEach((hunk) => {
    const tr = el("tr", "hunk");
    tr.append(el("td", "no"), el("td", "no"), el("td", "code", hunk.header));
    table.append(tr);
    hunk.lines.forEach((line) => {
      table.append(renderLine(line, fi, idx));
      idx += 1;
    });
  });
  const scroller = el("div", "diff-scroll");
  scroller.append(table);
  box.append(scroller);
  return box;
}

function renderLine(line, fi, idx) {
  const tr = el("tr", "line " + line.kind);
  tr.append(el("td", "no", line.old ?? ""), el("td", "no", line.new ?? ""));
  const code = el("td", "code");
  const plus = button("+", null, "plus");
  plus.title = "Add a comment; drag to select several lines";
  plus.onmousedown = (e) => {
    e.preventDefault();
    drag = { file: fi, start: idx, end: idx };
    highlight();
  };
  code.append(plus, el("span", "marker", MARK[line.kind]), el("span", "text", line.text));
  tr.append(code);
  tr.onmouseenter = () => {
    if (drag && drag.file === fi) {
      drag.end = idx;
      highlight();
    }
  };
  rows.set(fi + ":" + idx, tr);
  return tr;
}

// Sidebar: directories as collapsible groups, files that scroll to their diff.
function renderTree() {
  const root = { dirs: new Map(), files: [] };
  files.forEach((f, i) => {
    const parts = f.path.split("/");
    let node = root;
    parts.slice(0, -1).forEach((dir) => {
      if (!node.dirs.has(dir)) node.dirs.set(dir, { dirs: new Map(), files: [] });
      node = node.dirs.get(dir);
    });
    node.files.push({ name: parts.at(-1), index: i });
  });
  $("tree").textContent = "";
  $("tree").append(renderTreeNode(root));
}

function renderTreeNode(node) {
  const ul = el("ul", "tree-level");
  [...node.dirs.keys()].sort().forEach((name) => {
    const details = el("details");
    details.open = true;
    details.append(el("summary", "tree-dir", name), renderTreeNode(node.dirs.get(name)));
    const li = el("li");
    li.append(details);
    ul.append(li);
  });
  node.files.sort((a, b) => a.name.localeCompare(b.name)).forEach(({ name, index }) => {
    const li = el("li", "tree-file");
    li.append(button(name, () => reveal(index)));
    treeItems[index] = li;
    ul.append(li);
  });
  return ul;
}

function reveal(fi) {
  boxes[fi].classList.remove("collapsed");
  boxes[fi].scrollIntoView({ block: "start", behavior: "smooth" });
}

function filterFiles(query) {
  const q = query.trim().toLowerCase();
  files.forEach((f, i) => {
    const hidden = Boolean(q) && !f.path.toLowerCase().includes(q);
    boxes[i].hidden = hidden;
    treeItems[i].hidden = hidden;
  });
  document.querySelectorAll("#tree details").forEach((d) => {
    d.parentElement.hidden = !d.querySelector("li.tree-file:not([hidden])");
  });
}

function updateViewed() {
  $("review-count").textContent = viewed.size + " / " + files.length + " viewed";
}

document.addEventListener("mouseup", () => {
  if (!drag) return;
  const { file, start, end } = drag;
  drag = null;
  document.querySelectorAll("tr.selected").forEach((r) => r.classList.remove("selected"));
  openForm(file, Math.min(start, end), Math.max(start, end));
});

function highlight() {
  const [a, b] = [Math.min(drag.start, drag.end), Math.max(drag.start, drag.end)];
  document.querySelectorAll("tr.selected").forEach((r) => r.classList.remove("selected"));
  for (let i = a; i <= b; i++) rows.get(drag.file + ":" + i)?.classList.add("selected");
}

// Where the comments and forms of a line, or of the whole file, live.
function commentCell(fi, idx) {
  if (idx === FILE) return boxes[fi].querySelector(".file-comments");
  const line = rows.get(fi + ":" + idx);
  let row = line.nextElementSibling;
  if (!row || !row.classList.contains("comment-row")) {
    row = el("tr", "comment-row");
    const td = el("td");
    td.colSpan = 3;
    row.append(td);
    line.after(row);
  }
  return row.firstElementChild;
}

function pruneRow(fi, idx) {
  if (idx === FILE) return;
  const row = rows.get(fi + ":" + idx).nextElementSibling;
  if (row && row.classList.contains("comment-row") && !row.firstElementChild.childElementCount) row.remove();
}

function markCommented() {
  document.querySelectorAll("tr.commented").forEach((r) => r.classList.remove("commented"));
  comments.forEach((c) => {
    for (let i = c.start; i <= c.end; i++) rows.get(c.file + ":" + i)?.classList.add("commented");
  });
  const n = comments.length;
  $("pending-count").textContent = n ? n + (n === 1 ? " pending comment" : " pending comments") : "";
}

function where(fi, start, end) {
  if (start === FILE) return files[fi].path + " (file comment)";
  return files[fi].path + ":" + lineRef(files[fi].flat.slice(start, end + 1));
}

function openForm(fi, start, end, existing) {
  boxes[fi].classList.remove("collapsed");
  const form = el("div", "comment-form");
  const ta = el("textarea");
  ta.rows = 3;
  ta.placeholder = "Leave a comment";
  ta.value = existing ? existing.text : "";
  const cancel = () => {
    form.remove();
    if (existing) renderComment(existing);
    pruneRow(fi, end);
  };
  const submit = () => {
    const text = ta.value.trim();
    if (!text) return;
    form.remove();
    if (existing) {
      existing.text = text;
      renderComment(existing);
    } else {
      const c = { file: fi, start, end, text };
      comments.push(c);
      renderComment(c);
    }
    markCommented();
  };
  ta.onkeydown = (e) => {
    if (e.key === "Escape") { e.stopPropagation(); cancel(); }
    if (e.key === "Enter" && e.metaKey) { e.stopPropagation(); submit(); }
  };
  const actions = el("div", "form-actions");
  actions.append(button("Cancel", cancel), button(existing ? "Update comment" : "Add review comment", submit, "primary"));
  form.append(el("div", "where", where(fi, start, end)), ta, actions);
  commentCell(fi, end).append(form);
  ta.focus();
}

function renderComment(c) {
  const box = el("div", "comment");
  const head = el("div", "comment-head");
  head.append(el("span", "author", user), el("span", "tag pending", "Pending"));
  if (c.earlier) head.append(el("span", "tag earlier", "Earlier round"));
  head.append(el("span", "where", where(c.file, c.start, c.end)));
  const actions = el("div", "comment-actions");
  actions.append(
    button("Edit", () => { box.remove(); openForm(c.file, c.start, c.end, c); }),
    button("Delete", () => {
      box.remove();
      comments.splice(comments.indexOf(c), 1);
      pruneRow(c.file, c.end);
      markCommented();
    }),
  );
  box.append(head, el("div", "comment-body", c.text), actions);
  c.box = box;
  commentCell(c.file, c.end).append(box);
}

// "L12", "L12-L15", or old-file numbers for deleted lines only.
function lineRef(lines) {
  const news = lines.map((l) => l.new).filter((n) => n != null);
  if (news.length) return news[0] === news.at(-1) ? "L" + news[0] : "L" + news[0] + "-L" + news.at(-1);
  const olds = lines.map((l) => l.old);
  return "old L" + olds[0] + (olds.length > 1 ? "-L" + olds.at(-1) : "");
}

// The pending comments as text for the agent: where, the quoted lines, the note.
function reviewText() {
  return comments.map((c) => {
    if (c.start === FILE) return where(c.file, FILE, FILE) + "\n" + c.text;
    const lines = files[c.file].flat.slice(c.start, c.end + 1);
    const quoted = lines.map((l) => "> " + MARK[l.kind] + l.text).join("\n");
    return where(c.file, c.start, c.end) + "\n" + quoted + "\n" + c.text;
  }).join("\n\n");
}

File: src/ui/review.rs
//! The diff review, in the style of GitHub's "Files changed": a file tree
//! with a filter beside the files, each with its header, comments and diff.

use std::collections::{BTreeMap, HashSet};

use blit::{Input, PointerButton, Sides, Sizing, WidgetId};
use blit_desktop::atom::Rectangle;
use blit_desktop::layout::{flex, single, Align};
use blit_desktop::style::{Border, BorderRadius};
use blit_desktop::text::TextStyle;
use blit_desktop::widget::{scroll, text_input, TextInput};
use blit_desktop::{BoundsClip, Ui};

use crate::diff::{FileDiff, Kind, Status};
use crate::message::Scope;
use crate::state::{CommentAt, FileReview, Restored};
use crate::text::{self, Anchor};

use super::lines;
use super::theme;
use super::widgets::{self, panel, Button, Checkbox, Look, ScrollArea, Tag};

pub fn scope_label(scope: Scope) -> &'static str {
    match scope {
        Scope::Staged => "Staged changes only: plain git commit",
        Scope::Tracked => "Tracked files as they are: git commit -a",
        Scope::Worktree => "Working tree, untracked files included: git add runs first",
    }
}

pub(super) struct File {
    pub(super) diff: FileDiff,
    pub(super) viewed: bool,
    pub(super) collapsed: bool,
    /// Pending comments, sent with the decision.
    pub(super) comments: Vec<Comment>,
    /// Comments of an earlier attempt whose lines changed: not sent unless reopened.
    pub(super) outdated: Vec<Outdated>,
}

pub(super) struct Comment {
    pub(super) anchor: Anchor,
    pub(super) text: String,
    /// Carried over from an earlier attempt.
    pub(super) earlier: bool,
}

pub(super) struct Outdated {
    pub(super) quote: Vec<String>,
    pub(super) text: String,
}

/// The comment box being written; one at a time.
pub(super) struct Form {
    pub(super) file: usize,
    pub(super) anchor: Anchor,
    pub(super) text: String,
    pub(super) state: text_input::State,
    /// The index of the comment being edited, in its file.
    pub(super) editing: Option<usize>,
    /// Takes the keyboard focus once built.
    pub(super) focus: bool,
}

impl Form {
    pub(super) fn new(file: usize, anchor: Anchor, text: String, editing: Option<usize>) -> Self {
        let end = text.len();
        let state = text_input::State { cursor: end, anchor: end, offset_x: 0.0 };
        Form { file, anchor, text, state, editing, focus: true }
    }
}

/// Lines selected by dragging a "+" button, flat indices in one file.
#[derive(Clone, Copy)]
pub(super) struct Drag {
    pub(super) file: usize,
    pub(super) start: usize,
    pub(super) end: usize,
}

impl From<crate::Change> for File {
    fn from(change: crate::Change) -> Self {
        let mut file = File {
            collapsed: change.viewed,
            viewed: change.viewed,
            diff: change.diff,
            comments: Vec::new(),
            outdated: Vec::new(),
        };
        for restored in change.restored {
            match restored {
                Restored::Lines { start, end, text } => {
                    file.comments.push(Comment { anchor: Anchor::Lines { start, end }, text, earlier: true })
                }
                Restored::File { text } => file.comments.push(Comment { anchor: Anchor::File, text, earlier: true }),
                Restored::Outdated { quote, text } => file.outdated.push(Outdated { quote, text }),
            }
        }
        file
    }
}

#[derive(Default)]
pub struct Review {
    /// Read from git when the reviewer first opens the review.
    files: Option<Result<Vec<File>, String>>,
    tree: scroll::State,
    list: scroll::State,
    filter: String,
    filter_state: text_input::State,
    closed_dirs: HashSet<String>,
    drag: Option<Drag>,
    form: Option<Form>,
    /// A file picked in the tree, scrolled to once laid out.
    reveal: Option<usize>,
}

impl Review {
    pub fn open(&mut self, command: Option<&str>) {
        if self.files.is_none() {
            self.files = Some(crate::changes(command).map(|changes| changes.into_iter().map(File::from).collect()));
        }
    }

    fn loaded(&self) -> &[File] {
        match &self.files {
            Some(Ok(files)) => files,
            _ => &[],
        }
    }

    pub fn pending(&self) -> usize {
        self.loaded().iter().map(|file| file.comments.len()).sum()
    }

    pub fn viewed(&self) -> (usize, usize) {
        let files = self.loaded();
        (files.iter().filter(|file| file.viewed).count(), files.len())
    }

    /// The pending comments as the agent reads them.
    pub fn comment_texts(&self) -> Vec<String> {
        self.loaded()
            .iter()
            .flat_map(|file| file.comments.iter().map(|comment| text::comment(&file.diff, comment.anchor, &comment.text)))
            .collect()
    }

    /// What to keep for the next attempt; nothing when no diff was shown.
    pub fn reviews(&self) -> Option<(Vec<FileDiff>, Vec<FileReview>)> {
        let files = self.loaded();
        if files.is_empty() {
            return None;
        }
        let reviews = files
            .iter()
            .map(|file| FileReview {
                path: file.diff.path.clone(),
                viewed: file.viewed,
                comments: file
                    .comments
                    .iter()
                    .map(|comment| {
                        let (start, end) = match comment.anchor {
                            Anchor::File => (None, None),
                            Anchor::Lines { start, end } => (Some(start), Some(end)),
                        };
                        CommentAt { start, end, text: comment.text.clone() }
                    })
                    .collect(),
            })
            .collect();
        Some((files.iter().map(|file| file.diff.clone()).collect(), reviews))
    }
}

pub fn build(ui: Ui<'_>, review: &mut Review, user: &str, consumed: &mut bool) {
    let Review { files, tree, list, filter, filter_state, closed_dirs, drag, form, reveal } = review;
    let mut row = ui.layout(flex::row().gap(16.0));
    let files = match files {
        Some(Ok(files)) => files,
        Some(Err(error)) => {
            let message = format!("git error: {error}");
            let shown = widgets::wrapped(&message, theme::mono(theme::CODE), theme::colors().danger);
            row.child(flex::item().width(Sizing::grow())).insert(shown);
            return;
        }
        None => return,
    };
    // A drag over the "+" buttons ends where the pointer is released.
    if let (Some(selected), Input::PointerUp { button: PointerButton::Primary, .. }) = (*drag, *row.input()) {
        *drag = None;
        let anchor = Anchor::Lines { start: selected.start.min(selected.end), end: selected.start.max(selected.end) };
        *form = Some(Form::new(selected.file, anchor, String::new(), None));
    }
    let query = filter.trim().to_lowercase();
    let shown: Vec<bool> = files.iter().map(|file| file.diff.path.to_lowercase().contains(&query)).collect();

    let mut picked = None;
    row.child(flex::item().width(Sizing::fixed(240.0)).height(Sizing::grow())).build(|ui: Ui<'_>| {
        let mut pane = ui.layout(flex::column().gap(8.0));
        pane.child(flex::item()).build(|ui: Ui<'_>| {
            let mut field = ui.layout(single::layout().padding(Sides::xy(10.0, 6.0)));
            field.insert(panel(theme::colors().surface));
            let input = TextInput::new(filter_state, WidgetId::new("filter"), filter)
                .style(theme::sans(13.0))
                .color(theme::colors().text)
                .placeholder("Filter files...")
                .placeholder_color(theme::colors().muted)
                .selection_background(theme::colors().selected)
                .cursor_background(theme::colors().text);
            field.child(single::item().width(Sizing::grow())).build(input);
        });
        pane.child(flex::item().grow()).build(ScrollArea::new(tree, BoundsClip).build(|ui: Ui<'_>| {
            let mut entries = ui.layout(flex::column().gap(1.0));
            for entry in tree_entries(files, &shown, closed_dirs) {
                let (id, label) = match &entry {
                    Entry::Dir { path, name, depth, open } => {
                        (WidgetId::new(("tree dir", path)), format!("{}{}  {name}", indent(*depth), if *open { "▾" } else { "▸" }))
                    }
                    Entry::File { index, name, depth } => {
                        let tick = if files[*index].viewed { "  ✓" } else { "" };
                        (WidgetId::new(("tree file", *index)), format!("{}    {name}{tick}", indent(*depth)))
                    }
                };
                let button = Button::new(id, &label).look(Look::Quiet).style(theme::sans(13.0));
                if entries.child(flex::item()).build(button) {
                    match entry {
                        Entry::Dir { path, .. } => {
                            if !closed_dirs.remove(&path) {
                                closed_dirs.insert(path);
                            }
                        }
                        Entry::File { index, .. } => picked = Some(index),
                    }
                }
            }
        }));
    });

    if let Some(index) = picked {
        files[index].collapsed = false;
        *reveal = Some(index);
    }
    if let Some(index) = *reveal {
        let content = list.id.child("content");
        if let (Some(target), Some(top)) = (row.geometry(file_id(index)), row.geometry(content)) {
            list.scroll_to(target.y - top.y);
            *reveal = None;
        }
        row.request_frame();
    }

    // Rows are built only near the viewport, measured in content coordinates
    // from last frame's layout; a margin of one viewport keeps the next
    // screen ready while scrolling.
    let margin = list.viewport_extent.max(400.0);
    let (window_top, window_bottom) = (list.offset - margin, list.offset + list.viewport_extent + margin);
    let content_top = row.geometry(list.id.child("content")).map(|content| content.y);
    row.child(flex::item().grow()).build(ScrollArea::new(list, BoundsClip).build(|ui: Ui<'_>| {
        let mut column = ui.layout(flex::column().gap(GAP).padding(Sides::new().right(10.0)));
        if files.is_empty() {
            let note = widgets::text("(nothing to commit in this scope)", theme::sans(theme::BODY), theme::colors().muted);
            column.child(flex::item()).insert(note);
        }
        let mut estimate = 0.0;
        for (index, file) in files.iter_mut().enumerate() {
            if !shown[index] {
                continue;
            }
            let top = match (column.geometry(file_id(index)), content_top) {
                (Some(area), Some(content)) => area.y - content,
                _ => estimate,
            };
            estimate = top + lines::height_estimate(file) + GAP;
            let place = lines::Place { top, window_top, window_bottom };
            column
                .child(flex::item())
                .widget_id(file_id(index))
                .build(|ui: Ui<'_>| file_box(ui, index, file, drag, form, user, consumed, place));
        }
    }));
}

/// Space between file boxes.
const GAP: f32 = 14.0;

fn file_id(index: usize) -> WidgetId {
    WidgetId::new(("file", index))
}

fn indent(depth: usize) -> String {
    "    ".repeat(depth)
}

#[allow(clippy::too_many_arguments)]
fn file_box(
    ui: Ui<'_>,
    index: usize,
    file: &mut File,
    drag: &mut Option<Drag>,
    form: &mut Option<Form>,
    user: &str,
    consumed: &mut bool,
    place: lines::Place,
) {
    let mut boxed = ui.layout(flex::column());
    boxed.insert(Rectangle::new().border(Border::solid(1.0, theme::colors().border)).radius(BorderRadius::uniform(theme::RADIUS)));
    boxed
        .child(flex::item())
        .widget_id(lines::header_id(index))
        .build(|ui: Ui<'_>| header(ui, index, file, form));
    if !file.collapsed {
        boxed.child(flex::item()).build(|ui: Ui<'_>| lines::body(ui, index, file, drag, form, user, consumed, place));
    }
}

fn header(ui: Ui<'_>, index: usize, file: &mut File, form: &mut Option<Form>) {
    let mut row = ui.layout(flex::row().padding(Sides::xy(10.0, 6.0)).gap(10.0).align(Align::Center));
    let radius = if file.collapsed {
        BorderRadius::uniform(theme::RADIUS)
    } else {
        BorderRadius::new().top_left(theme::RADIUS).top_right(theme::RADIUS)
    };
    row.insert(Rectangle::new().background(theme::colors().surface).radius(radius));
    let chevron = if file.collapsed { "▸" } else { "▾" };
    if row.child(flex::item()).build(Button::new(WidgetId::new(("chevron", index)), chevron).look(Look::Quiet)) {
        file.collapsed = !file.collapsed;
    }
    let (status, color) = match file.diff.status {
        Status::Added => ("added", theme::colors().success),
        Status::Deleted => ("deleted", theme::colors().danger),
        Status::Renamed => ("renamed", theme::colors().purple),
        Status::Modified => ("modified", theme::colors().warning),
    };
    row.child(flex::item()).build(Tag { label: status, color });
    let path = match &file.diff.old_path {
        Some(old) => format!("{old} -> {}", file.diff.path),
        None => file.diff.path.clone(),
    };
    let bold = TextStyle { weight: 600, ..theme::mono(theme::CODE) };
    row.child(flex::item().width(Sizing::grow())).insert(widgets::text(&path, bold, theme::colors().text));
    let added = file.diff.lines().filter(|line| line.kind == Kind::Add).count();
    let removed = file.diff.lines().filter(|line| line.kind == Kind::Del).count();
    row.child(flex::item()).insert(widgets::text(&format!("+{added}"), theme::mono(theme::CODE), theme::colors().success));
    row.child(flex::item()).insert(widgets::text(&format!("-{removed}"), theme::mono(theme::CODE), theme::colors().danger));
    if row.child(flex::item()).build(Checkbox { id: WidgetId::new(("viewed", index)), label: "Viewed", checked: file.viewed }) {
        file.viewed = !file.viewed;
        file.collapsed = file.viewed;
    }
    let comment = Button::new(WidgetId::new(("file comment", index)), "Comment").look(Look::Quiet).style(theme::sans(theme::SMALL));
    if row.child(flex::item()).build(comment) {
        file.collapsed = false;
        *form = Some(Form::new(index, Anchor::File, String::new(), None));
    }
}

enum Entry {
    Dir { path: String, name: String, depth: usize, open: bool },
    File { index: usize, name: String, depth: usize },
}

#[derive(Default)]
struct Node {
    dirs: BTreeMap<String, Node>,
    files: Vec<(String, usize)>,
}

/// The tree rows for the shown files: directories first, then files, both
/// sorted, without the contents of closed directories.
fn tree_entries(files: &[File], shown: &[bool], closed: &HashSet<String>) -> Vec<Entry> {
    let mut root = Node::default();
    for (index, file) in files.iter().enumerate().filter(|(index, _)| shown[*index]) {
        let mut parts: Vec<&str> = file.diff.path.split('/').collect();
        let name = parts.pop().unwrap_or_default();
        let node = parts.iter().fold(&mut root, |node, dir| node.dirs.entry(dir.to_string()).or_default());
        node.files.push((name.to_string(), index));
    }
    let mut entries = Vec::new();
    walk(&root, "", 0, closed, &mut entries);
    entries
}

fn walk(node: &Node, prefix: &str, depth: usize, closed: &HashSet<String>, entries: &mut Vec<Entry>) {
    for (name, child) in &node.dirs {
        let path = format!("{prefix}{name}/");
        let open = !closed.contains(&path);
        entries.push(Entry::Dir { path: path.clone(), name: name.clone(), depth, open });
        if open {
            walk(child, &path, depth + 1, closed, entries);
        }
    }
    let mut files = node.files.clone();
    files.sort();
    entries.extend(files.into_iter().map(|(name, index)| Entry::File { index, name, depth }));
}
